setup.py

```diff
@@ -0,0 +1,149 @@
+
+import codecs
+import re
+from distutils.errors import DistutilsExecError
+from glob import glob
+from os.path import join, dirname, abspath
+from shutil import move, rmtree
+
+from setuptools import setup, find_packages
+SPHINX_AVAILABLE = False
+try:
+    from sphinx.setup_command import BuildDoc
+    SPHINX_AVAILABLE = True
+except ImportError:
+    pass
+
+here = abspath(dirname(__file__))
+
+
+def read(*parts):
+    with codecs.open(join(here, *parts), 'r') as fp:
+        return fp.read()
+
+
+def find_version(*file_paths):
+    version_file = read(*file_paths)
+    version_match = re.search(r"^__version__ = ['\"]([^'\"]*)['\"]",
+                              version_file, re.M)
+    if version_match:
+        return version_match.group(1)
+    raise RuntimeError("Unable to find version string.")
+
+
+cmdclass = {}
+if SPHINX_AVAILABLE:
+
+    class MakeDocs(BuildDoc):
+        """Port of old `cylc-make-docs`. Removed in #2989.
+
+        This class extends the Sphinx command class for setuptools. With
+        a difference that it tries to mimic the behaviour of the previous
+        `cylc-make-docs`.
+
+        So first it will execute `make-commands.sh`, which builds the
+        commands help information, in the appendices.
+
+        Then, instead of calling one builder, this class will call the
+        builder fot he single HTML, and also the builder for the multiple
+        HTML documentation.
+
+        Finally, one more tweak in this class is to move the doctrees
+        folder (in the same level as the documentation) to within the
+        documentation folder, named `.doctrees`, as before with
+        `cylc-make-docs`.
+        """
+
+        def run(self):  # type: () -> None
+            try:
+                self.spawn(["./doc/src/custom/make-commands.sh"])
+            except DistutilsExecError as e:
+                self.warn("Failed to run make-commands.sh")
+                raise e
+            self.do_run("html", "built-sphinx")
+            self.do_run("singlehtml", "built-sphinx-single")
+
+        def do_run(self, builder: str, output_dir: str):
+            """
+            Args:
+                builder (str): name of the Sphinx builder
+                output_dir (str): directory to write the documentation produced
+            """
+            self.builder = builder
+            self.builder_target_dirs = [
+                (builder, join(self.build_dir, output_dir))]
+            super().run()
+            # move doctrees to $build_dir/.doctrees
+            correct_doctrees = join(self.builder_target_dirs[0][1],
+                                    ".doctrees")
+            rmtree(correct_doctrees, ignore_errors=True)
+            move(self.doctree_dir, correct_doctrees)
+
+    cmdclass["build_sphinx"] = MakeDocs
+
+
+install_requires = [
+    'colorama==0.4.1',
+    'isodatetime==1!2.0.0',
+    'jinja2==2.10',
+    'markupsafe==1.1.1',
+    'python-jose==3.0.1',
+    'pyzmq==18.0.1'
```

Comment:
This means Cylc will *only* build with `18.0.1`?
When `18.0.2` is released this could cause users to downgrade for compatability?

If adding a version dependency I guess we should make it a minimum requirement not an absolute requirement (and through away the minor version unless necessary).

```suggestion
    'pyzmq>=18'
```

Jinja2 is an obvious exception as Cylc has tight coupling with the ever-changing templating language...

Reply:
All good for me. +1, will update tomorrow morning if there are no objections.

Reply:
:+1:

Reply:
Using `pyzmq==18.0.*`... `pyzmq>=18` means that if they released `pyzmq-19.?.?` our build would automatically pick it up :+1: